conformity/fields/email.py

```diff
@@ -0,0 +1,87 @@
+from __future__ import absolute_import, unicode_literals
+
+import re
+
+import attr
+
+from conformity.error import Error
+from conformity.fields.basic import UnicodeString
+from conformity.utils import strip_none
+from conformity.fields.net import IPAddress
+
+
+@attr.s
+class EmailAddress(UnicodeString):
+    """
+    On the shoulder of mighty Django (v2.0.x)
+    https://github.com/django/django/blob/stable/2.0.x/django/core/validators.py#L164
+    """
+    message = 'Enter a valid email address.'
+    code = 'invalid'
+    ip_schema = IPAddress()
+
+    user_regex = re.compile(
+        r"(^[-!#$%&'*+/=?^_`{}|~0-9A-Z]+(\.[-!#$%&'*+/=?^_`{}|~0-9A-Z]+)*\Z"  # dot-atom
+        r'|^"([\001-\010\013\014\016-\037!#-\[\]-\177]|\\[\001-\011\013\014\016-\177])*"\Z)',  # quoted-string
+        re.IGNORECASE
+    )
+    domain_regex = re.compile(
+        # max length for domain name labels is 63 characters per RFC 1034
+        r'((?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+)(?:[A-Z0-9-]{2,63}(?<!-))\Z',
+        re.IGNORECASE
+    )
+    literal_regex = re.compile(
+        # literal form, ipv4 or ipv6 address (SMTP 4.1.3)
+        r'\[([A-f0-9:\.]+)\]\Z',
+        re.IGNORECASE
+    )
+    domain_whitelist = ['localhost']
+
+    def __init__(self, message=None, code=None, whitelist=None):
+        if message is not None:
+            self.message = message
+        if code is not None:
+            self.code = code
+        if whitelist is not None:
+            self.domain_whitelist = whitelist
+
+    def errors(self, value):
+        # Get any basic type errors
+        result = super(EmailAddress, self).errors(value)
+        if result:
+            return result
+        if not value or '@' not in value:
+            return [Error('Not a valid email address (missing @ sign)')]
+
+        user_part, domain_part = value.rsplit('@', 1)
+        if not self.user_regex.match(user_part):
+            return [Error('Not a valid email address (invalid local user field)', user_part)]
+        if domain_part in self.domain_whitelist or self.is_domain_valid(domain_part):
+            return []
+        else:
+            try:
+                domain_part = domain_part.encode('idna').decode('ascii')
+                if self.is_domain_valid(domain_part):
+                    return []
+            except UnicodeError:
+                pass
+            return [Error('Not a valid email address (invalid domain field)', domain_part)]
+
+    def introspect(self):
+        return strip_none({
+            "type": "email_address",
+            "description": self.description,
+        })
+
+    def is_domain_valid(self, domain_part):
```

Comment:
I would move this method up to be above `introspect`, right after `errors`. Also, since it only uses static attributes, it could be a `@classmethod` with `cls` instead of `self` as the first argument.

Reply:
good point. edited.